ide/project.dependency/src/org/netbeans/modules/project/dependency/spi/ProjectArtifactsImplementation.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.project.dependency.spi;
+
+import java.util.Collection;
+import java.util.List;
+import javax.swing.event.ChangeListener;
+import org.netbeans.api.project.Project;
+import org.netbeans.modules.project.dependency.ArtifactSpec;
+import org.netbeans.modules.project.dependency.ProjectArtifactsQuery;
+
+/**
+ *
+ * @author sdedic
+ */
+public interface ProjectArtifactsImplementation {
+    /**
+     * Finds project artifact(s) that match the query. 
+     * @param query the query to answer
+     * @param previous result of previous buildsteps
+     * @param all
+     * @return 
+     */
+    public Result findArtifacts(ProjectArtifactsQuery.Filter query);
```

Comment:
I prefer [singletonizer](https://bits.netbeans.org/14/javadoc/org-netbeans-api-java-classpath/org/netbeans/spi/java/queries/BinaryForSourceQueryImplementation2.html) SPI over `ProjectArtifactsImplementation` and `Result` inner class.

Reply:
addressed in 70dcb7f24620